File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    
    unordered_map<string, unordered_set<string>> reported_user;
    unordered_map<string, int> count;
    
    // 보고된 사용자와 보고한 사용자 기록
    for (const auto& r : report) {
        size_t space_pos = r.find(' ');
        if (space_pos != string::npos) {
            string user_id = r.substr(0, space_pos);
            string reported_id = r.substr(space_pos + 1);
            reported_user[reported_id].insert(user_id);
        }
    }
    
    // 정지된 사용자에 대해 보고한 사용자에게 메일 발송
    for (const auto& [reported_id, user_set] : reported_user) {
        if (user_set.size() >= k) {
            for (const auto& uid : user_set) {
                count[uid]++;
            }
        }
    }
    
    // 결과 벡터 미리 할당 및 채우기
    vector<int> answer;
    answer.reserve(id_list.size());
    for (const auto& id : id_list) {
        answer.emplace_back(count[id]);
    }
    
    return answer;
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        HashMap<String, HashSet<String>> reportedUser = new HashMap<>();
        HashMap<String, Integer> count = new HashMap<>();
        
        for (String r : report) {
            String[] s = r.split(" ");
            String userId = s[0];
            String reportedId = s[1];
            
            if (!reportedUser.containsKey(reportedId)) {
                reportedUser.put(reportedId, new HashSet<>());
            }
            
            reportedUser.get(reportedId).add(userId);
        }
        
        for (Map.Entry<String, HashSet<String>> entry: reportedUser.entrySet()) {
            if (entry.getValue().size() >= k) {
                for (String uid : entry.getValue()) {
                    count.put(uid, count.getOrDefault(uid, 0) + 1);
                }
            }
        }
        
        int[] answer = new int[id_list.length];
        
        for (int i = 0; i < id_list.length; ++i) {
            answer[i] = count.getOrDefault(id_list[i], 0);
        }
        
        return answer;
    }
}